### ISSUE TITLE
Expose an elementary stream source which can be used by an app

### ISSUE THREAD

**Issue by kabhay** [google/ExoPlayer#7268]
### [REQUIRED] Use case description
Add support for MSE EME. Because certain apps may intend to use both MSE EME for certain formats along with the current model. Although the apps can directly use the decoders(decoder interfaces and implement the data injection, it would be good to have the Exo Player support this so that apps can use it even for MSE EME kind of playback. 

### Proposed solution
Add elementary stream data source which can inject demuxed audio and video packets. Support decryption of the same when they are found to be encrypted when the content is protected with different DRM schemes. 

### Alternatives considered
Not thought about it.

**Comment by kim-vde** [comment]
Could you explain in more details what would this enhancement be like? Do you suggest to add the possibility in ExoPlayer to provide elementary streams (instead of container data), with support for DRM data? What do you mean by MSE EME kind of playback?

**Comment by kabhay** [comment]
Hi Kim, 

Thanks for your quick reply. 

>>Could you explain in more details what would this enhancement be like? Do you suggest to add the >>possibility in ExoPlayer to provide elementary streams (instead of container data), with support for DRM data? 

In very simple words thats what it means. 

>>What do you mean by MSE EME kind of playback?

If you meant to share some document about MSE EME here is the wiki link for the same. This gives a brief overview of what MSE EME is all about. https://en.wikipedia.org/wiki/Media_Source_Extensions
But if you are looking for any further details etc. Please let me know. 

Warm Regards,
-Abhay.

**Comment by marcbaechinger** [comment]
MSE EME is for playing video in a browser. It specifies how a javascript application can provide media sources for being played by the media tag of the W3C spec. 

ExoPlayer is an Android media player to be used with Java/Kotlin. We do not plan to somehow expose ExoPlayer media sources to a webview or support MSE implementation inside ExoPlayer. I don't think this will happen any time soon. It's probably the same  for EME which is the spec for doing DRM with EME.

If you are looking for a JavaScript media player I recommend looking into [Shaka player](https://github.com/google/shaka-player).

We keep that open as enhancement, so you can check in later again.

**Comment by kabhay** [comment]
Although MSE EME is for playing media from a browser, my request here was to see if ExoPlayer can expose a elementary stream source which can be used by the applications handling the demuxing themselves. If you need any further information, please let me know.

**Comment by marcbaechinger** [comment]
Yeah, sorry my bad. I got confused by the Wikipedia article which was about MSE only not mentioning elementary streams. People of my team know about elementary streams and recognized my mistake. Sry about this, that's my fault so I reopened and you confirmed now. Thanks! I changed the title of the issue to reflect better that the enhancement is about elementary streams.

**Comment by kabhay** [comment]
Thank you Marc. My apologies too. Even I should have mentioned the subject as you have changed now. 
And probably given MSE EME implementors can still use ExoPlayer with elementary streams. Thanks again.
